File: app-server/src/main/java/com/github/arhor/simple/expense/tracker/data/repository/InternalUserRepository.java
package com.github.arhor.simple.expense.tracker.data.repository;

import java.util.Optional;

import org.springframework.data.jdbc.repository.query.Query;
import org.springframework.data.repository.CrudRepository;

import com.github.arhor.simple.expense.tracker.data.model.InternalUser;

public interface InternalUserRepository extends CrudRepository<InternalUser, Long> {

    Optional<InternalUser> findInternalUserByUsername(String username);

    @Query(name = "InternalUser.findByUsername")
    Optional<InternalUser.CompactProjection> findByUsername(String username);

    @Query(name = "InternalUser.findByExternalIdAndProvider")
    Optional<InternalUser.CompactProjection> findByExternalIdAndProvider(String externalId, String externalProvider);

    boolean existsByUsername(String username);

    boolean existsByExternalIdAndExternalProvider(String externalId, String externalProvider);
}

File: app-server/src/main/kotlin/com/github/arhor/simple/expense/tracker/data/repository/InternalUserRepository.kt
package com.github.arhor.simple.expense.tracker.data.repository

import com.github.arhor.simple.expense.tracker.data.model.InternalUser
import com.github.arhor.simple.expense.tracker.data.model.projection.CompactInternalUserProjection
import org.springframework.data.jdbc.repository.query.Query
import org.springframework.data.repository.CrudRepository

interface InternalUserRepository : CrudRepository<InternalUser, Long> {

    fun findInternalUserByUsername(username: String): InternalUser?

    @Query(name = "InternalUser.findByUsername")
    fun findByUsername(username: String): CompactInternalUserProjection?

    @Query(name = "InternalUser.findByExternalIdAndProvider")
    fun findByExternalIdAndProvider(externalId: String, externalProvider: String): CompactInternalUserProjection

    fun existsByUsername(username: String): Boolean

    fun existsByExternalIdAndExternalProvider(externalId: String, externalProvider: String): Boolean
}
